service/storage_service.cc

```diff
@@ -370,16 +370,37 @@ future<> storage_service::sync_raft_topology_nodes(mutable_token_metadata_ptr tm
         return *used_ips;
     };
 
+    using host_id_to_ip_map_t = std::unordered_map<locator::host_id, gms::inet_address>;
+    auto get_host_id_to_ip_map = [&, map = std::optional<host_id_to_ip_map_t>{}]() mutable -> future<const host_id_to_ip_map_t*> {
+        if (!map.has_value()) {
+            const auto ep_to_id_map = co_await _sys_ks.local().load_host_ids();
+            map.emplace();
+            map->reserve(ep_to_id_map.size());
+            for (const auto& [ep, id]: ep_to_id_map) {
+                const auto [it, inserted] = map->insert({id, ep});
+                if (!inserted) {
+                    on_internal_error(slogger, ::format("duplicate IP for host_id {}, first IP {}, second IP {}",
+                        id, it->second, ep));
+                }
+            }
+        }
+        co_return &*map;
+    };
+
     std::vector<future<>> sys_ks_futures;
 
+    auto remove_ip = [&](inet_address ip) -> future<> {
+        sys_ks_futures.push_back(_sys_ks.local().remove_endpoint(ip));
+
+        if (_gossiper.get_endpoint_state_ptr(ip) && !get_used_ips().contains(ip)) {
+            co_await _gossiper.force_remove_endpoint(ip, gms::null_permit_id);
```

Comment:
Haven't noticed this earlier, but is it OK to do it before / concurrently with `_sys_ks.local().remove_endpoint(ip));`?

Reply:
> Hmm, I think this function is bugged currently, and gossiper updates need to happen after corresponding system keyspace updates.

why?

Reply:
I'm not sure about remove_ip, the whole thing is hard to reason about..

but for example here
```cpp
            sys_ks_futures.push_back(_sys_ks.local().update_peer_info(*ip, info));
            if (!prev_normal.contains(id)) {
                co_await notify_joined(*ip);
            }
```

isn't it important that we put entry to peers before we send notifications? cc @gleb-cloudius 

Driver could receive such notification, then refresh peers but observe old state.

In general outside observes may see things happening out of order, and I'm worried it could cause hard to reproduce and debug bugs.


Or here
```cpp
            sys_ks_futures.push_back(_sys_ks.local().update_tokens(rs.ring.value().tokens));
            co_await _gossiper.add_local_application_state({
                { gms::application_state::TOKENS, gms::versioned_value::tokens(rs.ring.value().tokens) },
                { gms::application_state::CDC_GENERATION_ID, gms::versioned_value::cdc_generation_id(_topology_state_machine._topology.current_cdc_generation_id) },
                { gms::application_state::STATUS, gms::versioned_value::normal(rs.ring.value().tokens) }
            });
```

is it fine that other nodes might observe our gossip status update before we persist it locally? I think it's just asking for trouble...

Reply:
Moving the concurrency "outside" these lambdas would simplify reasoning a lot, put some order to things, while keeping the optimization intact. (There would still be concurrency between updates for different peers, which was the goal.)

Reply:
> The concurrency of system keyspace updates can be achieved by not co_awaiting on the process_normal_node etc. calls but putting them into a vector of futures. But inside process_normal_node etc., we should be doing operations in order.

It can't, this'll entail concurrent `token_metadata` mutations, e.g. multiple fibers mutating `_token_to_endpoint_map` through `update_normal_tokens`.

Reply:
Anyway, it's out of scope of this PR.